src/EventStore.Core/Services/Archiver/ArchiverService.cs

```diff
@@ -2,54 +2,188 @@
 // Event Store Ltd licenses this file to you under the Event Store License v2 (see LICENSE.md).
 
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using EventStore.Core.Bus;
+using EventStore.Core.Data;
+using EventStore.Core.Messages;
 using EventStore.Core.Services.Archiver.Storage;
-using EventStore.Plugins;
-using EventStore.Plugins.Licensing;
-using Microsoft.AspNetCore.Builder;
-using Microsoft.Extensions.Configuration;
-using Microsoft.Extensions.DependencyInjection;
-using Microsoft.Extensions.Logging;
-using ILogger = Serilog.ILogger;
+using EventStore.Core.Services.Archiver.Storage.Exceptions;
+using Serilog;
 
 namespace EventStore.Core.Services.Archiver;
 
-public class ArchiverService : IPlugableComponent {
-	protected static readonly ILogger Log = Serilog.Log.ForContext<ArchiverService>();
+public class ArchiverService :
+	IHandle<SystemMessage.ChunkLoaded>,
+	IHandle<SystemMessage.ChunkCompleted>,
+	IHandle<SystemMessage.ChunkSwitched>,
+	IHandle<ReplicationTrackingMessage.ReplicatedTo>,
+	IHandle<SystemMessage.SystemStart>,
+	IHandle<SystemMessage.BecomeShuttingDown>
+{
+	private static readonly ILogger Log = Serilog.Log.ForContext<ArchiverService>();
 
-	public string Name => "Archiver";
+	private readonly ISubscriber _mainBus;
+	private readonly IArchiveStorage _archiveStorage;
+	private readonly Queue<ChunkInfo> _uncommittedChunks;
+	private readonly ConcurrentDictionary<string, ChunkInfo> _existingChunks;
+	private readonly CancellationTokenSource _cts;
+	private readonly Channel<Commands.ArchiveChunk> _archiveChunkCommands;
 
-	public string DiagnosticsName => Name;
+	private readonly TimeSpan RetryInterval = TimeSpan.FromMinutes(1);
+	private long _replicationPosition;
 
-	public KeyValuePair<string, object>[] DiagnosticsTags => [];
+	public ArchiverService(ISubscriber mainBus, IArchiveStorage archiveStorage) {
+		_mainBus = mainBus;
+		_archiveStorage = archiveStorage;
 
-	public string Version => "0.0.1";
+		_uncommittedChunks = new();
+		_existingChunks = new();
+		_cts = new();
+		_archiveChunkCommands = Channel.CreateUnboundedPrioritized(
+			new UnboundedPrioritizedChannelOptions<Commands.ArchiveChunk> {
+				SingleWriter = true,
```

Comment:
```suggestion
				SingleWriter = false,
```
i think writes can come from the main queue but also from `ScheduleExistingChunksForArchiving`